bpf/kmesh/sendmsg.c

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2023 The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <linux/bpf.h>
+#include <bpf/bpf_endian.h>
+#include <bpf/bpf_helpers.h>
+#include "bpf_log.h"
+#include "sendmsg.h"
+
+/*
+ * sk msg is used to encode metadata into the payload when the client sends
+ * data to waypoint.
+ *
+ * The TLV format is used for encoding.
+ * Currently, only the dst ip and dst port of the original link need to be
+ * transferred to the waypoint.
+ *
+ * The current TLV format is simple. TLV structure is as follows:
+ * |TLV TYPE(1byte)|TLV LENGTH(4bytes)|TLV DATA|\
+ *
+ * When only the dst information needs to be transferred, the information
+ * added to the payload is as follows:
+ * |0x01|21|[dst ip]:[dst port]| ===> [dst ip] 4bytes, [:]1byte, [dst port] 4bytes
+ * |0xfe| 0|
+ * payload..
+ * total need add (1 + 4 + 4 + 1 + 4) + (1 + 4) = 19 bytes
```

Comment:
It seems [dst ip] and [dst port] is binary encoded, but we will include more metadata info, like pod name, pod ns...

So it's better to encode with json?